### PR TITLE
[Merged by Bors] - feat: integer operations in `norm_num`

### PR DESCRIPTION
**Pull request by digama0**
Another major rewrite of `norm_num`, this time to support returning integers as well as nats. Rational numbers are included but stubbed for now pending the rest of the algebraic typeclasses.

### CODE REVIEW

File: Mathlib/Data/Int/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+/-- Inductively define a function on `ℤ` by defining it at `b`, for the `succ` of a number greater
+than `b`, and the `pred` of a number less than `b`. -/
+@[elab_as_elim] protected def inductionOn' {C : ℤ → Sort _}
+    (z : ℤ) (b : ℤ) (H0 : C b) (Hs : ∀ k, b ≤ k → C k → C (k + 1))
+    (Hp : ∀ k ≤ b, C k → C (k - 1)) : C z := by
+  rw [← sub_add_cancel (G := ℤ) z b, add_comm]
+  exact match z - b with
+  | .ofNat n =>
+    let rec pos : ∀ n : ℕ, C (b + n)
```
Comment (semorrison) [inline Mathlib/Data/Int/Basic.lean:40]: Hmm, `docBlame` wants a doc-string here, but I think it's not possible to do inline.

Not sure if you prefer that `docBlame` changes, we work out how to write doc-strings inline here, or we just add doc-strings afterwards (seems unhelpful, these are only useful as code comments).

File: Mathlib/Algebra/GroupWithZero/Defs.lean [hunk 2]
```diff
@@ -75,19 +75,25 @@ class AddGroupWithOne (R : Type u) extends AddMonoidWithOne R, AddGroup R where
...
-def Int.cast [AddGroupWithOne R] : ℤ → R := AddGroupWithOne.intCast
+namespace Int
+
+@[coe] def cast [AddGroupWithOne R] : ℤ → R := AddGroupWithOne.intCast
 
-instance [AddGroupWithOne R] : CoeTail ℤ R where coe := Int.cast
+instance [AddGroupWithOne R] : CoeTail ℤ R where coe := cast
 
-theorem Int.cast_ofNat [AddGroupWithOne R] : (Int.cast (Int.ofNat n) : R) = Nat.cast n :=
+@[simp] theorem cast_ofNat [AddGroupWithOne R] : (cast (ofNat n) : R) = Nat.cast n :=
```
Comment (semorrison) [inline Mathlib/Algebra/GroupWithZero/Defs.lean:84]: The linter says this is redundant, but without it we get:
```
./././Mathlib/Data/Int/Cast/Defs.lean:56:24: error: unsolved goals
R : Type u_1
inst✝¹ inst✝ : AddGroupWithOne R
m n : ℕ
⊢ ↑↑(m + n) = ↑↑m + ↑↑n
```

**Comment by semorrison** [comment]
@digama0, I think the doc-string on `Result.toInt` actually belongs on the missing function `deriveInt`. I've added this function in https://github.com/leanprover-community/mathlib4/pull/507/commits/e5a9aa8292d65dfabab53b6b7998413e761a6a4b, but perhaps you can scrutinize this commit especially carefully to see that it is what you want.

File: Mathlib/Data/Int/Basic.lean (hunk 1 above)
Comment (digama0) [reply Mathlib/Data/Int/Basic.lean:40]: You can, it is in the really awkward place `let rec /-- doc -/ pos : ...`

**Comment by semorrison** [comment]
Looks great to me, and I'm already successfully using the additions in `abel1`.

bors d+

**Comment by digama0** [comment]
bors merge

**Pull request closed without merging**